File: python/src/dashgps/writers/geojsonw.py
"""GeoJSON writer with a hand-rolled emitter so key order is fixed. spec/21-outputs.md."""

from ..fmt import (
    P_ACCEL,
    P_ALT,
    P_ANGLE,
    P_DIST,
    P_LATLON,
    P_SPEED,
    fixed,
    iso_local,
    json_str,
)


def _coord(p):
    a = fixed(p.alt_m, P_ALT)
    base = "[" + fixed(p.lon, P_LATLON) + ", " + fixed(p.lat, P_LATLON)
    return base + ("]" if not a else ", " + a + "]")


def write(out, group, ctx):
    want_points = ctx.get("points", True)
    out.append('{\n  "type": "FeatureCollection",\n')
    out.append('  "properties": {\n')
    out.append('    "group": ' + json_str(group.label) + ",\n")
    out.append('    "generator": ' + json_str("dashgps " + ctx.get("version", "0.0.0")) + ",\n")
    out.append('    "time_is_naive": ' + ("true" if ctx.get("time_is_naive", True) else "false")
               + "\n")
    out.append("  },\n")
    out.append('  "features": [\n')

    first = True
    for ri, r in enumerate(group.runs):
        if r.glitch and not ctx.get("include_glitch"):
            continue
        if r.end - r.start < 2:
            continue
        if not first:
            out.append(",\n")
        first = False
        out.append("    {\n")
        out.append('      "type": "Feature",\n')
        out.append('      "properties": {\n')
        out.append('        "run": ' + str(ri) + ",\n")
        out.append('        "points": ' + str(r.end - r.start) + ",\n")
        out.append('        "distance_km": ' + (fixed(r.distance_km, P_DIST) or "0.000") + ",\n")
        out.append('        "outlier": ' + ("true" if r.glitch else "false") + ",\n")
        out.append('        "start": ' + json_str(iso_local(group.points[r.start].t)) + ",\n")
        out.append('        "end": ' + json_str(iso_local(group.points[r.end - 1].t)) + "\n")
        out.append("      },\n")
        out.append('      "geometry": {\n')
        out.append('        "type": "LineString",\n')
        out.append('        "coordinates": [\n')
        for i in range(r.start, r.end):
            out.append("          " + _coord(group.points[i]))
            out.append(",\n" if i < r.end - 1 else "\n")
        out.append("        ]\n      }\n    }")

    if want_points:
        for ri, r in enumerate(group.runs):
            if r.glitch and not ctx.get("include_glitch"):
                continue
            for i in range(r.start, r.end):
                p = group.points[i]
                if not first:
                    out.append(",\n")
                first = False
                out.append("    {\n")
                out.append('      "type": "Feature",\n')
                out.append('      "properties": {\n')
                out.append('        "timestamp": ' + json_str(iso_local(p.t)) + ",\n")
                out.append('        "speed_kmh": ' + (fixed(p.speed_kmh, P_SPEED) or "null")
                           + ",\n")
                out.append('        "heading_deg": ' + (fixed(p.heading_deg, P_ANGLE) or "null")
                           + ",\n")
                out.append('        "accel": [' + (fixed(p.ax, P_ACCEL) or "null") + ", "
                           + (fixed(p.ay, P_ACCEL) or "null") + ", "
                           + (fixed(p.az, P_ACCEL) or "null") + "],\n")
                out.append('        "run": ' + str(ri) + ",\n")
                out.append('        "outlier": ' + ("true" if p.outlier else "false") + "\n")
                out.append("      },\n")
                out.append('      "geometry": { "type": "Point", "coordinates": '
                           + _coord(p) + " }\n    }")
    out.append("\n  ]\n}\n")

File: js/src/writers/geojsonw.js
// GeoJSON writer with a hand-rolled emitter so key order is fixed. spec/21-outputs.md
// Mirrors python/src/dashgps/writers/geojsonw.py

import {
  P_ACCEL, P_ALT, P_ANGLE, P_DIST, P_LATLON, P_SPEED, fixed, isoLocal, jsonStr,
} from "../fmt.js";

function coord(p) {
  const a = fixed(p.altM, P_ALT);
  const base = "[" + fixed(p.lon, P_LATLON) + ", " + fixed(p.lat, P_LATLON);
  return a ? base + ", " + a + "]" : base + "]";
}

export function write(out, group, ctx) {
  const wantPoints = ctx.points !== false;
  out.push('{\n  "type": "FeatureCollection",\n');
  out.push('  "properties": {\n');
  out.push('    "group": ' + jsonStr(group.label) + ",\n");
  out.push('    "generator": ' + jsonStr("dashgps " + (ctx.version || "0.0.0")) + ",\n");
  out.push('    "time_is_naive": ' + (ctx.timeIsNaive !== false ? "true" : "false") + "\n");
  out.push("  },\n");
  out.push('  "features": [\n');

  let first = true;
  for (let ri = 0; ri < group.runs.length; ri++) {
    const r = group.runs[ri];
    if (r.glitch && !ctx.includeGlitch) continue;
    if (r.end - r.start < 2) continue;
    if (!first) out.push(",\n");
    first = false;
    out.push("    {\n");
    out.push('      "type": "Feature",\n');
    out.push('      "properties": {\n');
    out.push('        "run": ' + ri + ",\n");
    out.push('        "points": ' + (r.end - r.start) + ",\n");
    out.push('        "distance_km": ' + (fixed(r.distanceKm, P_DIST) || "0.000") + ",\n");
    out.push('        "outlier": ' + (r.glitch ? "true" : "false") + ",\n");
    out.push('        "start": ' + jsonStr(isoLocal(group.points[r.start].t)) + ",\n");
    out.push('        "end": ' + jsonStr(isoLocal(group.points[r.end - 1].t)) + "\n");
    out.push("      },\n");
    out.push('      "geometry": {\n');
    out.push('        "type": "LineString",\n');
    out.push('        "coordinates": [\n');
    for (let i = r.start; i < r.end; i++) {
      out.push("          " + coord(group.points[i]));
      out.push(i < r.end - 1 ? ",\n" : "\n");
    }
    out.push("        ]\n      }\n    }");
  }

  if (wantPoints) {
    for (let ri = 0; ri < group.runs.length; ri++) {
      const r = group.runs[ri];
      if (r.glitch && !ctx.includeGlitch) continue;
      for (let i = r.start; i < r.end; i++) {
        const p = group.points[i];
        if (!first) out.push(",\n");
        first = false;
        out.push("    {\n");
        out.push('      "type": "Feature",\n');
        out.push('      "properties": {\n');
        out.push('        "timestamp": ' + jsonStr(isoLocal(p.t)) + ",\n");
        out.push('        "speed_kmh": ' + (fixed(p.speedKmh, P_SPEED) || "null") + ",\n");
        out.push('        "heading_deg": ' + (fixed(p.headingDeg, P_ANGLE) || "null") + ",\n");
        out.push('        "accel": [' + (fixed(p.ax, P_ACCEL) || "null") + ", " +
                 (fixed(p.ay, P_ACCEL) || "null") + ", " +
                 (fixed(p.az, P_ACCEL) || "null") + "],\n");
        out.push('        "run": ' + ri + ",\n");
        out.push('        "outlier": ' + (p.outlier ? "true" : "false") + "\n");
        out.push("      },\n");
        out.push('      "geometry": { "type": "Point", "coordinates": ' + coord(p) + " }\n    }");
      }
    }
  }
  out.push("\n  ]\n}\n");
}
